### PR TITLE
Basic implementation of GCS store

### PR DESCRIPTION
**Pull request by fabxc**
@Bplotka 

The downloading of index files is working. The chunk reader is not yet implemented and tests are missing. But a first review round can't hurt.

### CODE REVIEW

File: cmd/thanos/example.go [hunk 1]
```diff
@@ -60,12 +53,6 @@ func registerExample(m map[string]setupFunc, app *kingpin.Application, name stri
 		}
 		g.AddGroup(sidecarGroup)
 
-		storeGroup, err := runStore(logger, metrics, *gcsBucket, *maxDiskCacheSize, *maxMemCacheSize)
```
Comment (Bplotka) [inline cmd/thanos/example.go]: This is temporary or we don't want store in example dir?

File: cmd/thanos/store.go [hunk 2]
```diff
@@ -37,20 +64,68 @@ func registerStore(m map[string]setupFunc, app *kingpin.Application, name string
...
+		g.Add(func() error {
+			return errors.Wrap(http.Serve(l, mux), "serve metrics")
+		}, func(error) {
+			l.Close()
+		})
 	}
-	defer gcsClient.Close()
 
-	// TODO(bplotka): Add store node logic for fetching blocks and exposing store API.
+	level.Info(logger).Log("msg", "I'm a store node")
```
Comment (Bplotka) [inline cmd/thanos/store.go]: `s/I'm a store node/Starting GCS store node` ?

File: pkg/store/gcs.go [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+	f, err := os.Create(filepath.Join(tmpdir, "index"))
+	if err != nil {
+		return errors.Wrap(err, "create local index copy")
+	}
+	r, err := indexObj.NewReader(ctx)
+	if err != nil {
+		return errors.Wrap(err, "create index object reader")
+	}
+	_, err = io.Copy(f, r)
```
Comment (Bplotka) [inline pkg/store/gcs.go]: not checked error?

Comment (Bplotka) [reply pkg/store/gcs.go]: OK it is checked down below. Do you mind renaming `err` to `copyErr` to make it verbose what is checked where?

File: pkg/store/gcs.go [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+		}
+		return errors.Wrap(err, "copy index file to disk")
+	}
+
+	if err := renameFile(tmpdir, bdir); err != nil {
+		return errors.Wrap(err, "rename block directory")
+	}
+
+	_, err = s.loadFromDisk(id)
+	return errors.Wrap(err, "load synced block")
```
Comment (Bplotka) [inline pkg/store/gcs.go]: shouldn't we wrap only on error?

File: pkg/store/gcs.go [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+	matchers, err := translateMatchers(req.Matchers)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	var g errgroup.Group
+	// TODO(fabxc): filter inspected blocks by labels.
+
+	for _, b := range s.blocks {
+		b := b
```
Comment (Bplotka) [inline pkg/store/gcs.go]: ?

File: pkg/store/gcs.go [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+		// where we consolidate requests.
+		g.Go(func() error {
+			set, err := tsdb.LookupChunkSeries(b.indexr, nil, matchers...)
+			if err != nil {
+				return errors.Wrap(err, "get series set")
+			}
+			for set.Next() {
+				lset, chunks, _ := set.At()
+
+				series := &storepb.Series{
```
Comment (Bplotka) [inline pkg/store/gcs.go]: We need to add these series into response, right? Or is that part of TODO?

File: cmd/thanos/example.go (hunk 1 above)
Comment (fabxc) [reply cmd/thanos/example.go]: I could've made it compile but it wouldn't have worked regardless. Figured I just drop it and we just revisit the command properly later.

File: cmd/thanos/store.go (hunk 2 above)
Comment (fabxc) [reply cmd/thanos/store.go]: Sure.

File: pkg/store/gcs.go (hunk 4 above)
Comment (fabxc) [reply pkg/store/gcs.go]: `errors.Wrap` returns `nil` again if the input error was `nil`. But I can make an explicit `if` if it seems to magical.

File: pkg/store/gcs.go (hunk 5 above)
Comment (fabxc) [reply pkg/store/gcs.go]: We start a goroutine below. If we don't do this, `b` will be overwritten.
It's the equivalent of 

```go
for _, b := range blocks {
    go func(b Block) {
        // ...
    }(b)
}
```

Just that we don't cal the func ourself with `errgroup`.

File: pkg/store/gcs.go (hunk 4 above)
Comment (Bplotka) [reply pkg/store/gcs.go]: Hm.. I would vote for explicit check (:

File: pkg/store/gcs.go (hunk 5 above)
Comment (Bplotka) [reply pkg/store/gcs.go]: yeah, standard loop + go routine/defer behavior... I guess you could change the resulting variable name to be bit different to be more explicit? (someone will miss that line while reading code and will look for `b` to check if you pack it into go routine or errgroup)

File: pkg/store/gcs_test.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+func TestGCSStore_downloadBlocks(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	project, ok := os.LookupEnv("GCP_PROJECT")
+	// TODO(fabxc): make it run against a mock store if no actual bucket is configured.
+	if !ok {
+		return
```
Comment (Bplotka) [inline pkg/store/gcs_test.go]: you can do t.Skip(...) first

File: pkg/store/gcs.go [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+	err := runutil.Repeat(60*time.Second, ctx.Done(), func() error {
+		if err := s.downloadBlocks(ctx); err != nil {
+			level.Warn(s.logger).Log("msg", "downloading missing blocks failed", "err", err)
+		}
+		if err := s.loadBlocks(); err != nil {
+			level.Warn(s.logger).Log("msg", "loading disk blocks failed", "err", err)
+		}
+		return nil
+	})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (fabxc) [reply pkg/store/gcs.go]: But then we do add something inside of the function to return, and will 90% forget to add the check back :)

File: pkg/store/gcs.go [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+
+// SyncBlocks synchronizes the stores state with the GCS bucket.
+func (s *GCSStore) SyncBlocks(ctx context.Context) {
+	// NOTE(fabxc): watches are not yet supported by the Go client library so we just
+	// do a periodic refresh.
+	err := runutil.Repeat(60*time.Second, ctx.Done(), func() error {
+		if err := s.downloadBlocks(ctx); err != nil {
+			level.Warn(s.logger).Log("msg", "downloading missing blocks failed", "err", err)
+		}
+		if err := s.loadBlocks(); err != nil {
```
[the remark this answers is not in the document]
Comment (fabxc) [reply pkg/store/gcs.go]: Yup, in download we just warn if downloading a particular block failed and continue. Only the iterator failed we return an error. But at that point we might have already downloaded a few blocks successfully which now should be loaded regardless.

File: pkg/store/gcs_test.go (hunk 7 above)
Comment (fabxc) [reply pkg/store/gcs_test.go]: Will do in a subsequent PR with another test.

File: pkg/store/gcs.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bplotka) [reply pkg/store/gcs.go]: xD

**Pull request merged**